explcheck/src/explcheck.lua

```diff
@@ -0,0 +1,168 @@
+local common = require("explcheck-common")
+local preprocessing = require("explcheck-preprocessing")
+-- local lexical_analysis = require("explcheck-lexical-analysis")
+-- local syntactic_analysis = require("explcheck-syntactic-analysis")
+-- local semantic_analysis = require("explcheck-semantic-analysis")
+-- local pseudo_flow_analysis = require("explcheck-pseudo-flow-analysis")
+
+-- Transform a singular into plural if the count is zero or greater than two.
+local function pluralize(singular, count)
+  if count == 1 then
+    return singular
+  else
+    return singular .. "s"
+  end
+end
+
+-- Colorize a string using ASCII color codes.
+local function colorize(text, ...)
+  local buffer = {}
+  for _, color_code in ipairs({...}) do
+    table.insert(buffer, "\27[")
+    table.insert(buffer, tostring(color_code))
+    table.insert(buffer, "m")
+  end
+  table.insert(buffer, text)
+  table.insert(buffer, "\27[0m")
+  return table.concat(buffer, "")
+end
+
+-- Convert a byte number in a file to a line and column number in a file.
+local function convert_byte_to_line_and_column(state, byte_number)
```

Comment:
Pass `line_starting_byte_numbers` directly to reduce the spread of the `state` god object. Also, compute it here rather in `explcheck-preprocessing.lua` to make the processing code do fewer menial tasks just for the CLI.

Reply:
Fixed in commit 506182f from this PR.